cirq/circuits/frozen_circuit_test.py

```diff
@@ -0,0 +1,2782 @@
+    ])
+
+    assert c + cirq.X(a) == cirq.FrozenCircuit(cirq.X(a))
+    assert c + [cirq.X(a)] == cirq.FrozenCircuit(cirq.X(a))
+    assert c + [[[cirq.X(a)], []]] == cirq.FrozenCircuit(cirq.X(a))
+    assert c + (cirq.X(a),) == cirq.FrozenCircuit(cirq.X(a))
+    assert c + (cirq.X(a) for _ in range(1)) == cirq.FrozenCircuit(cirq.X(a))
+    with pytest.raises(TypeError):
+        _ = c + cirq.X
+
+
+def test_radd_op_tree():
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+
+    c = cirq.FrozenCircuit()
+    assert [cirq.X(a), cirq.Y(b)] + c == cirq.FrozenCircuit([
+        cirq.Moment([cirq.X(a), cirq.Y(b)]),
+    ])
+
+    assert cirq.X(a) + c == cirq.FrozenCircuit(cirq.X(a))
+    assert [cirq.X(a)] + c == cirq.FrozenCircuit(cirq.X(a))
+    assert [[[cirq.X(a)], []]] + c == cirq.FrozenCircuit(cirq.X(a))
+    assert (cirq.X(a),) + c == cirq.FrozenCircuit(cirq.X(a))
+    assert (cirq.X(a) for _ in range(1)) + c == cirq.FrozenCircuit(cirq.X(a))
+    with pytest.raises(AttributeError):
+        _ = cirq.X + c
+    with pytest.raises(TypeError):
+        _ = 0 + c
+
+    # non-empty circuit addition
+    d = cirq.FrozenCircuit(cirq.Y(b))
+    assert [cirq.X(a)] + d == cirq.FrozenCircuit(
+        [cirq.Moment([cirq.X(a)]),
+         cirq.Moment([cirq.Y(b)])])
+    assert cirq.Moment([cirq.X(a)]) + d == cirq.FrozenCircuit(
+        [cirq.Moment([cirq.X(a)]),
+         cirq.Moment([cirq.Y(b)])])
+
+    # Preserves device.
+    c = cirq.FrozenCircuit(device=cirq.google.Bristlecone)
+    c2 = [] + c
+    assert c2.device is cirq.google.Bristlecone
+    assert c2 == c
+
+    # Validates versus device.
+    c = cirq.FrozenCircuit(device=cirq.google.Bristlecone)
+    with pytest.raises(ValueError, match='Unsupported qubit'):
+        _ = [cirq.X(cirq.NamedQubit('a'))] + c
+
+
+def test_bool():
+    assert not cirq.FrozenCircuit()
+    assert cirq.FrozenCircuit(cirq.X(cirq.NamedQubit('a')))
+
+
+def test_repr():
+    assert repr(cirq.FrozenCircuit()) == 'cirq.FrozenCircuit()'
+
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+    c = cirq.FrozenCircuit([
+        cirq.Moment([cirq.H(a), cirq.H(b)]),
+        cirq.Moment(),
+        cirq.Moment([cirq.CZ(a, b)]),
+    ])
+    cirq.testing.assert_equivalent_repr(c)
+    assert repr(c) == """cirq.FrozenCircuit([
+    cirq.Moment(
+        cirq.H(cirq.NamedQubit('a')),
+        cirq.H(cirq.NamedQubit('b')),
+    ),
+    cirq.Moment(),
+    cirq.Moment(
+        cirq.CZ(cirq.NamedQubit('a'), cirq.NamedQubit('b')),
+    ),
+])"""
+
+    c = cirq.FrozenCircuit(device=cg.Foxtail)
+    cirq.testing.assert_equivalent_repr(c)
+    assert repr(c) == 'cirq.FrozenCircuit(device=cirq.google.Foxtail)'
+
+    c = cirq.FrozenCircuit(cirq.Z(cirq.GridQubit(0, 0)), device=cg.Foxtail)
+    cirq.testing.assert_equivalent_repr(c)
+    assert repr(c) == """cirq.FrozenCircuit([
+    cirq.Moment(
+        cirq.Z(cirq.GridQubit(0, 0)),
+    ),
+], device=cirq.google.Foxtail)"""
+
+
+def test_empty_moments():
```

Comment:
This is almost entirely duplicated from `test_empty_moments` in `circuit_test.py`. It seems like a better way to handle this would be to write a single test that is parameterized over whether to use a frozen circuit, e.g.:
```python
@pytest.mark.parametrize('frozen', [False, True])
def test_empty_moments(frozen):
    ...
    circuit = cirq.Circuit(...)
    if frozen:
        circuit = circuit.freeze()

    cirq.testing.assert_has_diagram(...)
```
This same thing likely applies to much of the code below as well. It would be nice to avoid too much duplication so that going forward we have an easier time keeping tests in sync between Circuit and FrozenCircuit and don't miss testing common behavior on one or the other.

Reply:
This is true of most tests in frozen_circuit_test - there's a commit on this PR that literally copy-pastes the tests and replaces `Circuit` with `FrozenCircuit`. I agree that shared tests would be preferable - I'll apply this change.